File: app/src/main/java/com/dialogapp/dialog/ui/profilescreen/FollowingFragment.java
package com.dialogapp.dialog.ui.profilescreen;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.dialogapp.dialog.R;
import com.dialogapp.dialog.di.Injectable;
import com.dialogapp.dialog.model.AccountInfo;
import com.dialogapp.dialog.ui.base.BaseNetworkWatcherActivity;
import com.dialogapp.dialog.util.Resource;
import com.dialogapp.dialog.util.Status;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static com.dialogapp.dialog.ui.profilescreen.ProfileActivity.EXTRA_USERNAME;

public class FollowingFragment extends Fragment implements Injectable {
    @BindView(R.id.recycler_list)
    RecyclerView recyclerView;

    @BindView(R.id.swipeContainer)
    SwipeRefreshLayout swipeRefreshLayout;

    @BindView(R.id.empty_list)
    TextView emptyPlaceholder;

    @Inject
    ViewModelProvider.Factory viewModelFactory;

    private Unbinder unbinder;
    private FollowingRecyclerAdapter adapter;
    private String username;
    private final Observer<Resource<List<AccountInfo>>> observer = listResource -> {
        if (listResource != null)
            setData(listResource.status, listResource.data, listResource.message);
    };

    public static Fragment newInstance(String username) {
        FollowingFragment fragment = new FollowingFragment();
        Bundle bundle = new Bundle();
        bundle.putString(EXTRA_USERNAME, username);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            username = getArguments().getString(EXTRA_USERNAME);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.swipe_refresh_list_layout, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        adapter = new FollowingRecyclerAdapter(Glide.with(this));
        recyclerView.setLayoutManager(new LinearLayoutManager(this.getActivity(), LinearLayoutManager.VERTICAL, false));
        recyclerView.addItemDecoration(new DividerItemDecoration(this.getActivity(), DividerItemDecoration.VERTICAL));
        recyclerView.setAdapter(adapter);

        ProfileViewModel viewModel = ViewModelProviders.of(this, viewModelFactory).get(ProfileViewModel.class);
        viewModel.setFollowing(username);
        viewModel.getFollowingData().removeObserver(observer);
        viewModel.getFollowingData().observe(this, observer);

        swipeRefreshLayout.setOnRefreshListener(() -> {
            swipeRefreshLayout.setRefreshing(true);
            viewModel.refreshFollowingData();
        });

        ((BaseNetworkWatcherActivity) getActivity()).connectionViewModel.getConnectionStatus()
                .observe(getActivity(), isConnected -> {
                    if (isConnected != null && isConnected)
                        swipeRefreshLayout.setEnabled(true);
                    else
                        swipeRefreshLayout.setEnabled(false);
                });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    protected void setData(Status status, List<AccountInfo> data, String message) {
        // Set data ignoring the status
        if (data != null) {
            if (data.isEmpty()) {
                recyclerView.setVisibility(View.GONE);
                emptyPlaceholder.setVisibility(View.VISIBLE);
            } else {
                emptyPlaceholder.setVisibility(View.GONE);
                recyclerView.setVisibility(View.VISIBLE);

                adapter.submitList(data);
            }
        }

        swipeRefreshLayout.setRefreshing(status == Status.LOADING);
    }
}

File: app/src/main/java/com/dialogapp/dialog/ui/profile/following/FollowingFragment.kt
package com.dialogapp.dialog.ui.profile.following

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.dialogapp.dialog.GlideApp
import com.dialogapp.dialog.databinding.FragmentListBinding
import com.dialogapp.dialog.di.Injector
import com.dialogapp.dialog.model.EndpointData
import com.dialogapp.dialog.ui.base.BaseListFragment
import com.dialogapp.dialog.ui.profile.ProfileSharedViewModel
import com.dialogapp.dialog.ui.util.autoCleared
import com.dialogapp.dialog.vo.Status
import com.google.android.material.chip.Chip
import timber.log.Timber

class FollowingFragment : BaseListFragment() {

    private var binding by autoCleared<FragmentListBinding>()

    private lateinit var viewModel: FollowingViewModel
    private lateinit var profileSharedViewModel: ProfileSharedViewModel
    private lateinit var followingAdapter: FollowingAdapter

    companion object {
        private const val USERNAME = "username"
        private const val IS_SELF = "is_self"

        fun newInstance(username: String, isSelf: Boolean): FollowingFragment {
            val followingFragment = FollowingFragment()
            followingFragment.arguments = bundleOf(USERNAME to username, IS_SELF to isSelf)
            return followingFragment
        }
    }

    override fun onAttach(context: Context?) {
        viewModelFactory = Injector.get().viewModelFactory()
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        binding = FragmentListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val username = arguments?.getString(USERNAME)!!
        val isSelf = arguments?.getBoolean(IS_SELF)!!
        viewModel.showEndpoint(FollowingEndpointArgs(username, isSelf))
    }

    override fun initViewModel() {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(FollowingViewModel::class.java)
        profileSharedViewModel = ViewModelProviders.of(parentFragment!!).get(ProfileSharedViewModel::class.java)
    }

    override fun onSwipeRefresh() {
        viewModel.refresh()
    }

    override fun observe() {
        viewModel.endpointResult.observe(viewLifecycleOwner, Observer {
            val result = it ?: return@Observer

            binding.swipeRefresh.isRefreshing = result.status == Status.LOADING
            if (result.data != null) {
                followingAdapter.submitList(result.data)
            } else if (result.status == Status.ERROR) {
                followingAdapter.submitList(emptyList())
            }
        })
    }

    override fun onProfileClicked(username: String) {
        if (profileSharedViewModel.isProfileCurrentlyShown(username)) {
            Timber.d("Profile requested is currently on top of stack")
            return
        }
        super.onProfileClicked(username)
    }

    override fun initRecyclerAdapter() {
        val glide = GlideApp.with(this)
        followingAdapter = FollowingAdapter(glide, this)
        layoutManager = LinearLayoutManager(this.requireContext(), RecyclerView.VERTICAL, false)
        binding.recyclerPosts.layoutManager = layoutManager
        binding.recyclerPosts.adapter = followingAdapter
    }

    override fun getRecyclerView(): RecyclerView = binding.recyclerPosts

    override fun getSwipeRefreshLayout(): SwipeRefreshLayout = binding.swipeRefresh

    override fun getNotificationChip(): Chip = binding.chipNotification

    override fun setEndpointData(endpointData: EndpointData?) {

    }
}